### PR TITLE
T 2259640 table per template support

### PR DESCRIPTION
**Pull request by msgroi**

### CODE REVIEW

File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ */
+class FieldPrefixFunction {
+
+    private final String delimiter;
+
+    FieldPrefixFunction(String delimiter) {
+        this.delimiter = delimiter;
+    }
+
+    FieldValue apply(MTAmazonDynamoDBContextProvider mtContext, String tableIndex, String value) {
```
Comment (busjaeger) [inline src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java]: where is the code that turns a field value into a string?

File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MTAmazonDynamoDBBySharedTable.java [hunk 2]
```diff
@@ -219,74 +233,69 @@ public String toString() {
...
+            List<com.amazonaws.services.kinesis.model.Record> records = processRecordsInput.getRecords().stream()
+                    .map(RecordAdapter.class::cast).map(this::toMTRecord).collect(toList());
+            processor.processRecords(processRecordsInput.withRecords(records));
+        }
+
+        private com.amazonaws.services.kinesis.model.Record toMTRecord(RecordAdapter adapter) {
+            Record r = adapter.getInternalObject();
+            StreamRecord streamRecord = r.getDynamodb();
+            FieldValue fieldValue = new FieldPrefixFunction(".").reverse(streamRecord.getKeys().get(physicalTable.getPrimaryKey().getHashKey()).getS());
+            getMTContext().setContext(fieldValue.getMtContext());
```
Comment (busjaeger) [inline src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MTAmazonDynamoDBBySharedTable.java]: why set mt context here?

File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java (hunk 1 above)
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java]: One level up, in FieldMapper.apply() ... https://github.com/salesforce/mt-dynamo/blob/aae37cb2ec8a99b144e48aab35a2fa95b946b29b/src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldMapper.java#L37

File: src/main/java/com/salesforce/dynamodbv2/dynamodb/DynamoDBRetry.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+package com.salesforce.dynamodbv2.dynamodb;
+
+import com.amazonaws.services.dynamodbv2.model.LimitExceededException;
+
+import static java.lang.Math.pow;
+
+/*
+ * @author msgroi
+ */
+public class DynamoDBRetry {
```
[the remark this answers is not in the document]
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/dynamodb/DynamoDBRetry.java]: I had forgotten about that.  In this specific case, I really only want to retry when I get the LimitExceededException.  I suppose I could just ratchet up the general maxErrorRetry on the ClientConfiguration in my test.  Users of mt-dynamo will run into this problem and end up having to tweak the retry policy on their ClientConfiguration similarly.  I guess I could just put a 1-2 second delay after table create just to make it less likely to happen.

File: src/main/java/com/salesforce/dynamodbv2/mt/admin/AmazonDynamoDBAdminUtils.java [hunk 4]
```diff
@@ -36,35 +38,58 @@ public AmazonDynamoDBAdminUtils(AmazonDynamoDB amazonDynamoDB) {
...
+        } catch (TableInUseException e) {
+            if (e.getStatus().toUpperCase().equals("CREATING")) {
+                awaitTableActive(createTableRequest.getTableName(), pollIntervalSeconds, tableDDLOperationTimeoutSeconds);
+            } else {
+                throw new ResourceInUseException("table=" + e.getTableName() + " is in " + e.getStatus() + " status");
+            }
         }
     }
 
     public void deleteTableIfNotExists(String tableName, int pollIntervalSeconds, int timeoutSeconds) {
```
[the remark this answers is not in the document]
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/mt/admin/AmazonDynamoDBAdminUtils.java]: Fixed in f019397e3c4d76268b64ccd045e120ac7b59bc6c

File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MTAmazonDynamoDBBySharedTable.java (hunk 2 above)
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MTAmazonDynamoDBBySharedTable.java]: Unset context in a finally block ... 101ee1ad62730a68fb3999ce33fe5782e3c072db

FWIW, also fixed some formatting ... f2da4a6fda9e241c1963f8d1710e86753e338f1c

File: src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java (hunk 1 above)
Comment (msgroi) [reply src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/FieldPrefixFunction.java]: fixed some formatting ... f2da4a6fda9e241c1963f8d1710e86753e338f1c

**Pull request merged**